src/math/src/sha512.cairo

```diff
@@ -103,25 +103,48 @@ pub fn fpow(mut base: u128, mut power: u128) -> u128 {
     // Return invalid input error
     assert!(base != 0, "fpow: invalid input");
 
-    let mut base_u128: u256 = base.into();
-    let mut result: u256 = 1;
+    let mut result = 1;
     while (power != 0) {
-        if power % 2 != 0 {
-            result = (result * base_u128);
+        let (q, r) = DivRem::div_rem(power, 2);
+        if r == 1 {
+            result = (result * base);
         }
-        base_u128 = (base_u128 * base_u128);
-        power = power / 2;
+        base = base * base;
+        power = q;
     };
 
-    result.try_into().unwrap()
+    result
+}
+
+pub fn two_pow(mut power: u128) -> u128 {
+    let two_squarings: Array<u128> = array![
+        0x2, 0x4, 0x10, 0x100, 0x10000, 0x100000000, 0x10000000000000000
+    ];
+
+    let mut i = 0;
+    let mut result = 1;
+    while (power != 0) {
+        let (q, r) = DivRem::div_rem(power, 2);
+        if r == 1 {
+            result = result * *two_squarings[i];
+        }
+        i = i + 1;
+        power = q;
+    };
+
+    result
 }
 
 fn math_shl(x: u128, n: u128) -> u128 {
-    x * fpow(2, n) % BoundedInt::max()
+    x * two_pow(n) % BoundedInt::max()
```

Comment:
`x` and `n` are among the worst variable names a human could choose.
Can you give them more expressive names?

Reply:
Yeah, not too expressive but maybe it makes sense in context of the document the implementation follows?